skynet/dgpu/__init__.py

```diff
@@ -1,41 +1,40 @@
-#!/usr/bin/python
+import logging
+from contextlib import asynccontextmanager as acm
 
 import trio
+import urwid
 
-from hypercorn.config import Config
-from hypercorn.trio import serve
-from quart_trio import QuartTrio as Quart
+from skynet.config import Config
+from skynet.dgpu.tui import init_tui
+from skynet.dgpu.daemon import dgpu_serve_forever
+from skynet.dgpu.network import NetConnector
 
-from skynet.dgpu.compute import SkynetMM
-from skynet.dgpu.daemon import SkynetDGPUDaemon
-from skynet.dgpu.network import SkynetGPUConnector
 
+@acm
+async def open_worker(config: Config):
+    # suppress logs from httpx (logs url + status after every query)
+    logging.getLogger("httpx").setLevel(logging.WARNING)
 
-async def open_dgpu_node(config: dict) -> None:
-    '''
-    Open a top level "GPU mgmt daemon", keep the
-    `SkynetDGPUDaemon._snap: dict[str, list|dict]` table
-    and *maybe* serve a `hypercorn` web API.
+    tui = None
+    if config.tui:
+        tui = init_tui(config)
 
-    '''
-    conn = SkynetGPUConnector(config)
-    mm = SkynetMM(config)
-    daemon = SkynetDGPUDaemon(mm, conn, config)
+    conn = NetConnector(config)
 
-    api: Quart|None = None
-    if 'api_bind' in config:
-        api_conf = Config()
-        api_conf.bind = [config['api_bind']]
-        api: Quart = await daemon.generate_api()
+    try:
+        n: trio.Nursery
+        async with trio.open_nursery() as n:
+            if tui:
+                n.start_soon(tui.run)
 
-    tn: trio.Nursery
-    async with trio.open_nursery() as tn:
-        tn.start_soon(daemon.snap_updater_task)
+            n.start_soon(conn.iter_poll_update, config.poll_time)
 
-        # TODO, consider a more explicit `as hypercorn_serve`
-        # to clarify?
-        if api:
-            tn.start_soon(serve, api, api_conf)
+            yield conn
 
-        # block until cancelled
-        await daemon.serve_forever()
+    except *urwid.ExitMainLoop:
+        ...
+
+
+async def _dgpu_main(config: Config):
```

Comment:
abs love this composition and how it immediately indicates the supervision hierarchy and daemon-blocking semantics 👍🏼